src/misc.cpp

```diff
@@ -221,6 +225,19 @@ const std::string compiler_info() {
   return compiler;
 }
 
+/// large_page_alloc() tries to allocate aligned large page memory on systems
```

Comment:
This comment is a bit misleading. There is no small/large page memory on Linux, it's all unified. The difference is the page table entry descriptors: with large pages, the descriptors define 2-MB virtual-to-physical mappings instead of 4-kB mappings. This means that when using large-page PTEs, the number of PTEs is much smaller. The smaller number relieves TLB (translation look-aside buffer) pressure, and it makes the MMU pagewalker to hit the CPU caches more frequently when handling TLB misses and looking for the PTEs.

It's the madvise() call below that converts 4-kB PTEs to 2-MB PTEs. This is why the alignment has to be 2 MB on alloc so that the whole buffer range can use the larger PTEs.

Reply:
Also, don't mix this with the huge page pools ( /sys/kernel/mm/hugepages/ ). That is actually a different thing than transparent huge pages. When allocating from those pools, it is done with mmap() using MAP_HUGE_2MB or MAP_HUGE_1GB flags. So with the fixed pools, you're actually allocating 2MB or 1GB paged memory.

Transparent huge pages are best-effort (but it's pretty good effort). Using MAP_HUGE_* is black-and-white: you either get them or you don't. I have a patch for that, too, but it's pretty intrusive, and it requires user setup. I'm not considern sending PRs of that work at the moment ( https://github.com/skiminki/Stockfish/commit/8a31cefa06807b04c1f321e35728ab217ca97ff3 ).

Side note: 1GB page size gives me around +1% boost over 2MB page size, so it's not worth the effort to get it upstreamed, I suppose.